zerver/data_import/import_util.py

```diff
@@ -730,8 +728,18 @@ def process_emojis(
 
         emoji_records.append(emoji_record)
 
-    # Run downloads in parallel
-    run_parallel_wrapper(partial(get_emojis, emoji_dir), upload_emoji_list, threads=threads)
+        # Directly download the emoji and patch the file_name with the correct extension
+        # based on the content-type returned by the server. This is needed because Slack
+        # sometimes returns an emoji url with .png extension despite the file being a gif.
+        content_type = get_emojis(emoji_dir, emoji_url, emoji_path)
+        if content_type is not None:
+            file_extension = guess_extension(content_type)
+            if file_extension is None:
+                continue
+            old_file_name = emoji_record["file_name"]
+            new_file_name = f"{old_file_name.rsplit('.', 1)[0]}{file_extension}"
```

Comment:
Why do we need to correct this?  We rename the image when we upload anyways, and if we have a content-type, then the extension gets ignored (IIRC).

Reply:
@alexmv Hmm, i'm not sure to which place in the code you're referring to regarding the renaming of the image during upload; but in any case, I think it's not in these import codepaths. The extension in import doesn't get ignored - the final file path is generated based on the `realm_emoji["file_name"]`, so inherits the extension:

https://github.com/zulip/zulip/blob/083163605045661085a12ff87d9bd865728fb44b/zerver/lib/import_realm.py#L878-L883

An extra thing we could consider here would be to also add `content_type` field to the `emoji_record` dicts (they become entries in `emoji/records.json` in the export data) so that it can be used in my `process_emojis` function I'm adding in this PR instead of doing the `guess_type(record["file_name"])[0] ` thing.

But either way, the `file_name` extensions have to be fixed up here correctly.

Reply:
Oh.  Hm.  We don't have the luxury of renaming the filename to `get_emoji_file_name` to hash it, because we've already used the previous filename in the message content?  That's unfortunate, since it walks around our attempts to make the filenames not easily guessable by clients.

Given that, I guess this is the best we can do.

Reply:
>  We don't have the luxury of renaming the filename to get_emoji_file_name to hash it, because we've already used the previous filename in the message content? 

This statement is incorrect -- we render the message content at message import time, which is after emoji are imported.  We can merge this as-is, and do a follow-up PR to hash the filenames on import so that they're not predictable. For Zulip-generated emoji, we hash the AVATAR_SALT with the id; since we don't know the id here yet, we can hash the AVATAR_SALT with the emoji name, since that's unique within the realm.  It doesn't need to be the _same_ algorithm everywhere, as long as the output is not enumerable or guessable by a client.